keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+with ResizePolicy set to NoRestart for all its Containers.
+
+1. Initiating actor updates Pod's Container.ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements (e.g. limits are not below
+   requested resources, QoS class does not change, ResourceQuota not exceeded..).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers reject the update, API Server responds with an
+     appropriate error message.
+1. API Server updates PodSpec object with the new desired ResourceRequirements.
+1. Kubelet observes that Pod's Container.ResourceRequirements and
+   ContainerStatus.ResourcesAllocated differ. It checks its Node allocatable
+   resources to determine if the new desired ResourceRequirements fit the Node.
+   * _Case 1_: Kubelet finds new ResourceRequirements fit. It sets
+     ResizingPod.Status to true, and ResizingPod.Reason to InProgress, then
+     applies resized cgroup limits to the Pod and its Containers, and once
+     successfully done, updates Pod's ContainerStatus.ResourcesAllocated to
+     reflect the desired ResourceRequirements. It then sets ResizingPod.Status
+     to false, and ResizingPod.Reason to Success.
+     - If at the same time, a new Pod was scheduled against the capacity used by
+       this resource resize, that Pod is rejected during Kubelet admission if
+       Node has no more room.
+   * _Case 2_: Kubelet finds that desired ResourceRequirements does not fit.
+     - Kubelet checks to see if evicting lower priority Pods on the Node can
+       successfully resize the Pod. If yes, it sets ResizingPod.Status to true,
+       ResizingPod.Reason to PreEmpting, and initiates pre-emption of lower
+       priority Pods via Eviction API. Once lower priority Pods have been
+       evicted, it sets ResizingPod.Reason to InProgress, and the flow
+       continues as above.
+     - If Kubelet is unable to create enough room by evicting lower priority
+       Pods, it sets ResizingPod.Reason to FailedNodeCapacity, and
+       ResizingPod.Status to false.
+1. Scheduler, in parallel, observes that Container.ResourceRequirements and
+   ContainerStatus.ResourcesAllocated differ, and uses
+   max(ResourceRequirements, ResourcesAllocated) when computing resources
+   available on the Node.
+   * This can temporarily result in sum of Pod resources for the Node exceeding
+     Node's allocatable resources if a Pod was scheduled to that Node in parallel.
+1. Scheduler observes ResizingPod.Reason and/or ContainerStatus.ResourcesAllocated
+   fields have changed.
+   * _Case 1_: ResizingPod.Status is false, and ResizingPod.Reason is Success,
+     and Pod's Container.ResourceRequirements matches
+     ContainerStatus.ResourcesAllocated. Scheduler updates its cache with resized
+     ContainerStatus.ResourcesAllocated values.
+   * _Case 2_: ResizingPod.Status is false, and ResizingPod.Reason is
```

Comment:
i think this case should be removed.  a kubelet should just keep trying to converge to the desired size, reclaiming memory is not immediate, so its not a point in time decision that the kubelet makes but one it converges towards.

Reply:
This case 2 is just to explicitly call out what would happen on the Scheduler if Kubelet hits case 2 above. (In this flow, Scheduler simply continues to observe while Kubelet converges, nothing eventful happens)

Reply:
This case has been removed in updated design. Resolving this issue.